power_grid_model_c/power_grid_model/include/power_grid_model/main_core/sparse_idx_vector.hpp

```diff
@@ -0,0 +1,50 @@
+// SPDX-FileCopyrightText: 2022 Contributors to the Power Grid Model project <dynamic.grid.calculation@alliander.com>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_MAIN_CORE_TOPOLOGY_INDEX_HPP
+#define POWER_GRID_MODEL_MAIN_CORE_TOPOLOGY_INDEX_HPP
+
+#include "../power_grid_model.hpp"
+
+#include <boost/range.hpp>
+
+namespace power_grid_model::main_core {
+
+class SparseIdxVector {
```

Comment:
I thought SparseIdxVector in place of (eg. source_bus_indptr)
And SparseDataVector in Mathmodel's source param. (Actual data of source)

Is that logic okay?

Reply:
@mgovers  I shall merge the two. One class with two vector members. one of indptr and data, both internal to it. 
And externally we only retrieve the data as we see fit.

Reply:
Apologies, I misinterpreted what you wanted to do initially. But I also see what you implemented now and it looks good to me. (see below what I misunderstood)

## New implementation with only a `SparseVector`

This is your current implementation (commit [e08ff3b](https://github.com/PowerGridModel/power-grid-model/pull/379/commits/e08ff3bca7f70e31d4d64fdb2ca1ad55094b1933)).

As per how it will be used, maybe i wasn't concrete enough on what the usage would look like. i would suggest is something like:

```cpp
struct SourceParam {
  double a{};
  double b{};
};

template <typename T> class SparseVector {
  private:
    IdxVector indptr;
    std::vector<T> data;

  public:
    // vector interface
    // map interface
}

struct CalculationParams {
  SparseVector<SourceParam> source_param;
};
```

The advantage is that there's no `SparseIdxVector` exposed to the outside world (range-based over index-based for loops). The downside is that the data storage is row-based, not columnar, as for the current implementation, although I am not convinced yet that that is a problem.

I now also understand why you initially went with your alternative approach. See below.

## Your initial approach

The main advantage of your initial approach (2 data types with no explicit relation between the two) is that the data is kept separate from the indices - as with the current existing implementation. you can also keep the columnar data storage format.

The downside of your original approach is that your map-like interface returns `std::pair`, which the data should not have to care about. I would expect in your approach that the `SparseIdxVector` returns ONLY an `Idx`, so that your `SparseDataVector` can just be the `std::vector`

## `SparseVector` with columnar storage

To implement the new implementation with columnar storage format would require variadic templates (that would almost become something like the `Container` class, but it would be near-impossible to deal with duplicate data types like two fields that are both `double`) or something like your original implementation

```cpp
class SparseIdxVector {
  // blabla
};

template <typename T> class SparseDataVector{
  public:
    SparseDataVector(SparseIdxVector const& indices) : indices_{indices} {};
    // or with std::shared_ptr<SparseIdxVector const>

    // vector-like interface
    // map-like interface

  private:
    SparseIdxVector const& indices_;
    // or with std::shared_ptr<SparseIdxVector const>

    std::vector<T> data_;
};
```

otherwise you would end up with a lot of data duplication (a `SparseIdxVector` per `SparseDataVector`). Your initial approach but with only `IdxVector` specified would be far superior.